lib/wkhtmltopdf_pre_extract.rb

```diff
@@ -0,0 +1,57 @@
+###
+# wkhtmltopdf_binary_gem Copyright 2013 The University of Iowa
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License. You may obtain a copy
+# of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+require 'rbconfig'
+require 'zlib'
+
+suffix = case RbConfig::CONFIG['host_os']
+         when /linux/
+           os = `. /etc/os-release 2> /dev/null && echo ${ID}_${VERSION_ID}`.strip
+
+           os_based_on_ubuntu_16_04 = os.start_with?('ubuntu_16.') || os.start_with?('ubuntu_17.')
+           os = 'ubuntu_16.04' if os_based_on_ubuntu_16_04
+
+           os_based_on_ubuntu_18_04 = os.start_with?('ubuntu_18.') || os.start_with?('ubuntu_19.') || os.start_with?('elementary') || os.start_with?('linuxmint') || os.start_with?('pop') || os.start_with?('zorin')
+           os = 'ubuntu_18.04' if os_based_on_ubuntu_18_04
+
+           os_based_on_ubuntu_20_04 = os.start_with?('ubuntu_20.')
+           os = 'ubuntu_20.04' if os_based_on_ubuntu_20_04
+
+           os_based_on_centos_6 = (os.empty? && File.read('/etc/centos-release').start_with?('CentOS release 6')) || (os.start_with?('amzn_') && !os.start_with?('amzn_2'))
+           os = 'centos_6' if os_based_on_centos_6
+
+           os_based_on_centos_7 = os.start_with?('amzn_2')
+           os = 'centos_7' if os_based_on_centos_7
+
+           os_based_on_debian_9 = (/(debian_9|deepin)/ =~ os)
+           os = 'debian_9' if os_based_on_debian_9
+
+           os_based_on_debian = !os_based_on_debian_9 && os.start_with?('debian')
+           os = 'debian_10' if os_based_on_debian
+
+           os_based_on_archlinux = os.start_with?('arch_') || os.start_with?('manjaro_')
+           os = 'archlinux' if os_based_on_archlinux
+
+           architecture = RbConfig::CONFIG['host_cpu'] == 'x86_64' ? 'amd64' : 'i386'
+
+           "#{os}_#{architecture}"
+         when /darwin/
+           'macos_cocoa'
+         else
+           'unknown'
+         end
+
+suffix = ENV['WKHTMLTOPDF_HOST_SUFFIX'] unless ENV['WKHTMLTOPDF_HOST_SUFFIX'].to_s.empty?
+
+binary = "./bin/wkhtmltopdf_#{suffix}"
+
+if File.exist?("#{binary}.gz") && !File.exist?(binary)
```

Comment:
I think this result depends on the current working directory, which means the path is only correct if this is executed from the gem's root directory. But we'd ultimately be running this with the current working directory set to an application's root directory, so the path wouldn't be correct.

The way to ensure we're referring to the correct path independent of the current working directory is to use `File.expand_path` with a path relative to `__dir__`.

That said, I'm not sure we should bother fixing this. After thinking about the problem some more, I don't think this is the best approach. While I'd mentioned the problem was similar to what we ran into with the `Webdrivers` gem, I think there's a subtle but significant difference: That gem is only used in the test environment, so calling the appropriate setup code "ahead of time" in a test helper was a good fit. On the other hand, this gem is used in all environments, so I think we should aim for a more general solution (that's not limited to the test environment). I'm experimenting with an alternative approach of unzipping the binary during gem installation. That would solve it for any environment _and_ we wouldn't have to add special purpose code in our application.

Reply:
I've opened https://github.com/mycase/wkhtmltopdf-binary/pull/4 with that approach.